onnxruntime/core/providers/cpu/math/einsum_auxiliary_ops.cc

```diff
@@ -0,0 +1,296 @@
+  return true;
+}
+
+// Parse diagnoal elements along the 2 innermost dimensions
+// eg: input_shape = [1, 2, 3, 3]
+
+// This implementation provides flexibility as to which of the 2 innermost dim values is preserved 
+// via `preserve_innermost_dim_val` param
+
+// preserve_innermost_dim_val == true,
+//       output_shape = [1, 2, 1, 3] => the diagonal contains 3 elements and the dim value of the innermost dim is preserved
+
+// preserve_innermost_dim_val == false,
+//       output_shape = [1, 2, 3, 1] => the diagonal contains 3 elements and the dim value of the non-innermost dim is preserved
+static Tensor DiagonalInnermostDims(const Tensor& input, bool preserve_innermost_dim_val, AllocatorPtr allocator) {
+  const char* input_data = reinterpret_cast<const char*>(input.DataRaw());
+  const auto& input_dims = input.Shape().GetDims();
+  auto rank = input_dims.size();
+  const size_t element_size_in_bytes = input.DataType()->Size();
+
+  // This is an internal method and we already have finished all validations in the calling method.
+  // We proceed without duplicating all validations again here.
+
+  // We have a minimalistic check here to make sure the innermost dims have the same dim value
+  // as the calling method may have done a transpose before calling this method
+  ORT_ENFORCE(input_dims[rank - 2] == input_dims[rank - 1],
+              "The innermost dims should have the same dim value to parse the diagonal elements");
+
+  std::vector<int64_t> output_dims;
+  output_dims.reserve(rank);
+
+  int64_t num_iterations = 1;  // Flatten the outermost dims - this will be the number of iterations
+  for (size_t i = 0; i < rank - 2; ++i) {
+    auto input_dim_value = input_dims[i];
+    num_iterations *= input_dim_value;
+    output_dims.push_back(input_dim_value);
+  }
+
+  if (preserve_innermost_dim_val) {
+    output_dims.push_back(1);
+    output_dims.push_back(input_dims[rank - 1]);
+  } else {
+    output_dims.push_back(input_dims[rank - 1]);
+    output_dims.push_back(1);
+  }
+
+  int64_t inner_stride = input_dims[rank - 1];        // offset to move over the innermost dim
+  int64_t base_stride = inner_stride * inner_stride;  // offset to move over all the axes except the 2 innermost dims
+
+  // Pass in allocator as that will be used as an allocator deleter by the framework
+  // and it will de-allocate the memory for this intermediate tensor when it goes out of scope
+  Tensor output(input.DataType(), output_dims, allocator);
+  char* output_data = reinterpret_cast<char*>(output.MutableDataRaw());
+
+  int64_t output_iter = 0;
+  // TODO: Parallelize this operation
+  for (int64_t i = 0; i < num_iterations; ++i) {
+    auto base_offset = i * base_stride;
+    for (int64_t j = 0; j < inner_stride; ++j) {
+      memcpy(output_data + output_iter * element_size_in_bytes,
+             input_data + (base_offset + j * inner_stride + j) * element_size_in_bytes,
+             element_size_in_bytes);
+      output_iter++;
```

Comment:
If the memcpy call doesn't get optimized out in a release build may want to templatize this loop on the element size and just do an assignment.

Reply:
Done.